internal/dataplane/parser/translators/httproute.go

```diff
@@ -34,100 +43,301 @@ func TranslateHTTPRoute(route *gatewayv1beta1.HTTPRoute) []*HTTPRouteTranslation
+	seenHeaders := make(map[string]struct{})
+	headers := make([]gatewayv1beta1.HTTPHeaderMatch, 0, len(m.Match.Headers))
+	for _, header := range m.Match.Headers {
+		name := strings.ToLower(string(header.Name))
+		header.Name = gatewayv1beta1.HTTPHeaderName(strings.ToLower(string(header.Name)))
+		if _, ok := seenHeaders[name]; ok {
+			continue
 		}
-		backendKey.WriteString(".")
+		seenHeaders[name] = struct{}{}
+		headers = append(headers, header)
+	}
+	sort.Slice(headers, func(i, j int) bool {
+		return headers[i].Name < headers[j].Name
+	})
 
-		if backendRef.Namespace != nil {
-			backendKey.WriteString(string(*backendRef.Namespace))
+	// According to the spec of HTTPQueryParamMatch:
+	//
+	// If multiple entries specify equivalent query param names, only the first
+	// entry with an equivalent name MUST be considered for a match. Subsequent
+	// entries with an equivalent query param name MUST be ignored.
+	//
+	seenQueryParams := make(map[string]struct{})
+	queryParams := make([]gatewayv1beta1.HTTPQueryParamMatch, 0, len(m.Match.QueryParams))
+	for _, queryParam := range m.Match.QueryParams {
+		if _, ok := seenQueryParams[queryParam.Name]; ok {
+			continue
 		}
-		backendKey.WriteString(".")
+		seenQueryParams[queryParam.Name] = struct{}{}
+		queryParams = append(queryParams, queryParam)
+	}
 
-		backendKey.WriteString(string(backendRef.Name))
-		backendKey.WriteString(".")
+	keySource := struct {
+		Method  *gatewayv1beta1.HTTPMethod
+		Headers []gatewayv1beta1.HTTPHeaderMatch
+		Query   []gatewayv1beta1.HTTPQueryParamMatch
+	}{
+		m.Match.Method,
+		headers,
+		queryParams,
+	}
 
-		if backendRef.Port != nil {
-			backendKey.WriteString(strconv.Itoa(int(*backendRef.Port)))
-		}
-		backendKey.WriteString(".")
+	return mustMarshalJSON(keySource)
+}
 
-		if backendRef.Weight != nil {
-			backendKey.WriteString(strconv.Itoa(int(*backendRef.Weight)))
-		}
+// getHTTPBackendRefsKey computes a key from a list of backendRefs.
+// The order of backedRefs is not important.
+func getHTTPBackendRefsKey(backendRefs ...gatewayv1beta1.HTTPBackendRef) string {
+	backendKeys := make([]string, 0, len(backendRefs))
 
-		backendKeys = append(backendKeys, backendKey.String())
+	// create a list of backend keys
+	for _, backendRef := range backendRefs {
+		backendKeys = append(backendKeys, mustMarshalJSON(backendRef))
 	}
 	sort.Strings(backendKeys)
 
-	return httpBackendRefsKey(strings.Join(backendKeys, ";"))
+	return strings.Join(backendKeys, ";")
+}
+
+// getFiltersKey computes a key from a list of filters.
+// The order of the filters is not important.
+func getFiltersKey(filters []gatewayv1beta1.HTTPRouteFilter) string {
+	filterKeys := make([]string, 0, len(filters))
+
+	for _, filter := range filters {
+		filterKeys = append(filterKeys, mustMarshalJSON(filter))
+	}
+	sort.Strings(filterKeys)
+
+	return strings.Join(filterKeys, ";")
+}
+
+func mustMarshalJSON[T any](val T) string {
+	key, err := json.Marshal(val)
+	if err != nil {
+		return ""
+	}
```

Comment:
Why do we want the empty strings? 🤔

Reply:
what else are you proposing here? Returning an error makes this function not needed at all. But the errors are not propagated in translator. Anyway, if `json.Marshal` failed, the input ins very wrong.

Reply:
The intention behind my comment was that I'm not certain yet what would be the effect of empty values in the concatenated string that's built from the return values in `getFiltersKey()` and `getHTTPBackendRefsKey()`.

Reply:
That should never happen. To trigger that you would need to pass an argument unsupported type:
> // Channel, complex, and function values cannot be encoded in JSON.
> // Attempting to encode such a value causes Marshal to return
> // an UnsupportedTypeError.
> //
https://github.com/golang/go/blob/master/src/encoding/json/encode.go#L150-L152

Or unported value, like [NaN or Inf](https://github.com/golang/go/blob/master/src/encoding/json/encode.go#L575-L577) , [cyclic dependencies with large structs](https://github.com/golang/go/blob/master/src/encoding/json/encode.go#L785-L788).

What we use for the key does not qualify to trigger that branch.

Reply:
Ok got it, in such case I'd suggest to add a comment as that's very implicit IMHO

Reply:
Added function comment, ptal

Reply:
The stringbuilder approach instead of JSON marshaling is faster. That makes sense, as we got rid of the reflection in JSON from marshaling. It lowers the performance hit versus no routes consolidation. I'm thinking of building the keys manually.